In [1]:
import os
import tables as tb

os.environ["POLARS_MAX_THREADS"] = "128"
import polars as pl
import polars.selectors as cs

In [2]:
%load_ext watermark
%watermark -vp polars,tables

Python implementation: CPython
Python version       : 3.11.3
IPython version      : 8.13.2

polars: 0.20.6
tables: 3.8.0



# Acquiring a candidate set of proteins for structural searches
To search for capsid folds, we used foldseek. Instead of directly predicting a structure (which can be computationally expensive), we used the ProstT5 model implemented in foldseek that translates amino acid alphabets (proteins) to a structural alphabet. This can use pre-existing text alignment strategies (notably implemented in mmseqs2, reimplemented for foldseek) to search structural alphabet databases instead of the most computationally expensive 3D structural alignment.

For our investigations, rather than focus on all proteins in our dataset (which would take a long time), we only kept proteins belonging to protein clusters that had either VOG-detectable capsids and unannotated proteins. Our implicit assumption is that these represent protein clusters that are structurally similar, which we indirectly infer by detecting if these unannotated proteins contain capsid-like structural folds.

In [3]:
# for simplicity we are changing the "vog_category"
# for major capsid proteins to "MCP" instead of "structural"
mcp_query = (
    pl.col("vog_annot").str.contains("(?i)(Major )?capsid protein") &
    (~pl.col("vog_annot").str.contains("Minor")) &
    pl.col("vog_bitscore").ge(75)
)

ptn_info = (
    pl.scan_csv("supplementary_table_2.tsv", separator="\t")
    .filter(pl.col("dataset") == "test")
    .with_columns(
        vog_category = (
            pl.when(mcp_query)
            .then(pl.lit("MCP"))
            .otherwise(pl.col("vog_category"))
        )
    )
    .drop(cs.starts_with("phrog"))
    .drop("genome", "genome_id", "dataset")
    .collect()
)



ptn_info.head()

shape: (5, 5)
┌───────────────────────────────────┬────────┬──────────────┬──────────────────┬──────────────┐
│ ptn                               ┆ ptn_id ┆ vog_bitscore ┆ vog_annot        ┆ vog_category │
│ ---                               ┆ ---    ┆ ---          ┆ ---              ┆ ---          │
│ str                               ┆ i64    ┆ f64          ┆ str              ┆ str          │
╞═══════════════════════════════════╪════════╪══════════════╪══════════════════╪══════════════╡
│ IMGVR_UViG_2561511195_000001|256… ┆ 0      ┆ null         ┆ unknown function ┆ unknown      │
│ IMGVR_UViG_2561511195_000001|256… ┆ 1      ┆ null         ┆ unknown function ┆ unknown      │
│ IMGVR_UViG_2561511195_000001|256… ┆ 2      ┆ null         ┆ unknown function ┆ unknown      │
│ IMGVR_UViG_2561511195_000001|256… ┆ 3      ┆ null         ┆ unknown function ┆ unknown      │
│ IMGVR_UViG_2561511195_000001|256… ┆ 4      ┆ null         ┆ unknown function ┆ unknown      │
└───────────────────────────────────┴────────┴──────────────┴──────────────────┴──────────────┘

We select our candidates from all combinations of protein-genome clustering. We only consider genome clustering with the hyperparameters of k=15 and resolution=1.0 (high):

In [4]:
candidates: list[pl.LazyFrame] = []

ptn_embedding_methods = ["esm-small", "esm-large", "pst-small", "pst-large", "genslm"]
with tb.open_file("embedding-based_protein_clusters_per_genome_cluster.h5") as fp:
    for gm_node in fp.root:
        genome_method = gm_node._v_name

        for ptn_method in ptn_embedding_methods:
            clustering_metadata = gm_node["metadata"][:]
            ptn_clusters = gm_node[ptn_method][:]
            genome_clusters = gm_node["genome"][:]

            for pclu, gclu, (genome_k, genome_res, protein_k, protein_res) in zip(
                ptn_clusters, genome_clusters, clustering_metadata
            ):
                if (genome_k != 15) or (genome_res != 1.0) or (protein_res == 0.85):
                    continue

                ptn_df = (
                    ptn_info
                    .lazy()
                    .with_columns(
                        genome_cluster = pl.lit(gclu),
                        protein_cluster = pl.lit(pclu),
                    )
                    .with_columns(
                        cluster_size = pl.len().over("genome_cluster", "protein_cluster")
                    )
                    .filter(pl.col("cluster_size") > 1)
                    .with_columns(
                        num_categories = pl.n_unique("vog_category").over("genome_cluster", "protein_cluster"),
                        interested = (
                            pl.col("vog_category")
                            .is_in(["MCP", "unknown"])
                            .all()
                            .over("genome_cluster", "protein_cluster")
                        )
                    )
                    .filter(
                        (pl.col("num_categories") == 2)
                        & pl.col("interested")
                    )
                    .select("ptn")
                )

                candidates.append(ptn_df)

candidate_proteins = (
    pl.concat(pl.collect_all(candidates))
    ["ptn"]
    .value_counts()
    .filter(pl.col("count") >= 10)
)

len(candidate_proteins)

72322

To this set of 72,322 proteins, we considered that these proteins belong to sequence identity-based clusters, since those are the most similar proteins in the traditional sense. These additional proteins are homologs to those above. Think of these more as like positive controls.

In [5]:
mmseqs_clusters = (
    pl.read_csv(
        "sequence_identity_clusters.tsv", 
        separator="\t",
        has_header=False,
        new_columns=["rep", "ptn"],
    )
    .with_columns(
        cluster_size = pl.col("ptn").len().over("rep"),
        cluster_id = pl.col("rep").rle_id(),
    )
)

First we need to detect the mmseqs clusters that contain capsids:

In [6]:
detectable_mcp_clusters = (
    ptn_info
    .filter(pl.col("vog_category") == "MCP")
    .select("ptn")
    .join(mmseqs_clusters, on="ptn")
    .filter(pl.col("cluster_size") > 1)
    ["cluster_id"]
)

len(detectable_mcp_clusters)

46564

Then choose all mmseqs clusters that contain both capsids and unannotated proteins:

In [7]:
unknown_with_mcp_clusters = (
    ptn_info
    .join(mmseqs_clusters, on="ptn")
    .filter(pl.col("cluster_id").is_in(detectable_mcp_clusters))
    .group_by("cluster_id")
    .agg(
        pl.first("cluster_size"),
        cats = pl.col("vog_category").unique(),
    )
    # should also have at least 1 MCP due to the filter on line 4
    .filter(pl.col("cats").list.contains("unknown"))
    ["cluster_id"]
)

mmseqs_ptn_candidates = (
    mmseqs_clusters
    .filter(pl.col("cluster_id").is_in(unknown_with_mcp_clusters))
    ["ptn"]
)

len(mmseqs_ptn_candidates)

40357

Thus, the total number of candidate proteins to search for structural similarity to capsid proteins is:

In [8]:
total_candidates = set(candidate_proteins["ptn"]) | set(mmseqs_ptn_candidates)

len(total_candidates)

100704

We then extracted the 100,704 candidated MCPs into a separate FASTA file and ran the following `foldseek` (v9.427df8a) commands to:
1. Convert these protein sequences in to the 3Di-structure sequence:

```bash
# download ProstT5 model
foldseek databases ProstT5 prostt5 tmp

# translate protein amino acids to structure
foldseek createdb FASTAFILE structDB --prostt5-model prostt5
```

2. Search against Protein Data Bank structures

```bash
# use foldseek to download structures
foldseek databases PDB pdb tmp

# then search
foldseek search structDB pdb/pdb searchDB tmp
```

3. Then compute alignment stats

```bash
foldseek convertalis structDB pdb/pdb searchDB
```

The structural alignments from this analysis are available with the supplementary data at the path: `fig4/capsid_candidates_structural_search_against_pdb.tsv`

Then, to determine which PDB structures corresponded to capsids, we queried PDB at the web server ([https://www.rcsb.org](https://www.rcsb.org)) with: `capsid, major capsid, coat, minor capsid, virion`, and selected all PDB IDs the came up. This list of capsid PDB IDs is available with supplementary data at: `fig4/capsid_pdb_ids.txt`.

# Analyzing structural searches

We used the above 2 files to process the structural alignment results:

In [9]:
foldseek_columns = "query,target,fident,alnlen,mismatch,gapopen,qstart,qend,tstart,tend,evalue,bits,qlen,tlen".split(",")

with open("supplementary_data/fig4/capsid_pdb_ids.txt") as fp:
    capsid_pdb_ids = {line.rstrip() for line in fp}

struct_aln = (
    pl.scan_csv(
        "supplementary_data/fig4/capsid_candidates_structural_search_against_pdb.tsv",
        separator="\t",
        has_header=False,
        new_columns=foldseek_columns,
    )
    .filter(pl.col("bits") >= 100.)
    .with_columns(
        pdb_id = pl.col("target").str.extract("^(.*)?-.*$").str.to_uppercase(),
        chain = pl.col("target").str.split("_").list.get(1).str.to_uppercase(),
    )
    .with_columns(
        is_capsid = pl.col("pdb_id").is_in(capsid_pdb_ids),
    )
    .sort("bits", descending=True)
    # choose best PDB hit for each query protein
    .unique(subset="query", keep="first")
    .collect()
)

struct_aln

shape: (69_546, 17)
┌────────────────┬────────────────┬────────┬────────┬───┬──────┬───────────────┬───────┬───────────┐
│ query          ┆ target         ┆ fident ┆ alnlen ┆ … ┆ tlen ┆ pdb_id        ┆ chain ┆ is_capsid │
│ ---            ┆ ---            ┆ ---    ┆ ---    ┆   ┆ ---  ┆ ---           ┆ ---   ┆ ---       │
│ str            ┆ str            ┆ f64    ┆ i64    ┆   ┆ i64  ┆ str           ┆ str   ┆ bool      │
╞════════════════╪════════════════╪════════╪════════╪═══╪══════╪═══════════════╪═══════╪═══════════╡
│ IMGVR_UViG_GVM ┆ 8h2i-assembly1 ┆ 0.306  ┆ 516    ┆ … ┆ 517  ┆ 8H2I          ┆ BU    ┆ true      │
│ AG-S-ERX555967 ┆ .cif.gz_bu     ┆        ┆        ┆   ┆      ┆               ┆       ┆           │
│ -93_…          ┆                ┆        ┆        ┆   ┆      ┆               ┆       ┆           │
│ IMGVR_UViG_330 ┆ 2fs3-assembly1 ┆ 0.196  ┆ 287    ┆ … ┆ 258  ┆ 2FS3          ┆ F     ┆ true      │
│ 0034061_001882 ┆ .cif.gz_F      ┆        ┆        ┆   ┆      ┆               ┆       ┆           │
│ |330…          ┆                ┆        ┆        ┆   ┆      ┆               ┆       ┆           │
│ IMGVR_UViG_GVM ┆ 8h2i-assembly1 ┆ 0.342  ┆ 540    ┆ … ┆ 436  ┆ 8H2I-ASSEMBLY ┆ AB-5  ┆ false     │
│ AG-M-330000224 ┆ .cif.gz_ab-5   ┆        ┆        ┆   ┆      ┆ 1.CIF.GZ_AB   ┆       ┆           │
│ 2-2_…          ┆                ┆        ┆        ┆   ┆      ┆               ┆       ┆           │
│ IMGVR_UViG_330 ┆ 6ibc-assembly1 ┆ 0.166  ┆ 365    ┆ … ┆ 374  ┆ 6IBC-ASSEMBLY ┆ A-60  ┆ false     │
│ 0037453_000373 ┆ .cif.gz_A-60   ┆        ┆        ┆   ┆      ┆ 1.CIF.GZ_A    ┆       ┆           │
│ |330…          ┆                ┆        ┆        ┆   ┆      ┆               ┆       ┆           │
│ IMGVR_UViG_330 ┆ 1if0-assembly1 ┆ 0.134  ┆ 299    ┆ … ┆ 256  ┆ 1IF0          ┆ D     ┆ true      │
│ 0036985_003246 ┆ .cif.gz_D      ┆        ┆        ┆   ┆      ┆               ┆       ┆           │
│ |330…          ┆                ┆        ┆        ┆   ┆      ┆               ┆       ┆           │
│ …              ┆ …              ┆ …      ┆ …      ┆ … ┆ …    ┆ …             ┆ …     ┆ …         │
│ IMGVR_UViG_639 ┆ 3bjq-assembly1 ┆ 0.2    ┆ 323    ┆ … ┆ 303  ┆ 3BJQ          ┆ B     ┆ false     │
│ 857001_000001| ┆ .cif.gz_B      ┆        ┆        ┆   ┆      ┆               ┆       ┆           │
│ 6398…          ┆                ┆        ┆        ┆   ┆      ┆               ┆       ┆           │
│ IMGVR_UViG_GVM ┆ 8h2i-assembly1 ┆ 0.43   ┆ 65     ┆ … ┆ 436  ┆ 8H2I-ASSEMBLY ┆ AB-5  ┆ false     │
│ AG-M-330000427 ┆ .cif.gz_ab-5   ┆        ┆        ┆   ┆      ┆ 1.CIF.GZ_AB   ┆       ┆           │
│ 7-10…          ┆                ┆        ┆        ┆   ┆      ┆               ┆       ┆           │
│ IMGVR_UViG_330 ┆ 7mh2-assembly1 ┆ 0.143  ┆ 290    ┆ … ┆ 279  ┆ 7MH2          ┆ B     ┆ false     │
│ 0020047_017371 ┆ .cif.gz_B      ┆        ┆        ┆   ┆      ┆               ┆       ┆           │
│ |330…          ┆                ┆        ┆        ┆   ┆      ┆               ┆       ┆           │
│ IMGVR_UViG_330 ┆ 8ppv-assembly1 ┆ 0.153  ┆ 277    ┆ … ┆ 447  ┆ 8PPV          ┆ A     ┆ false     │
│ 0042449_002081 ┆ .cif.gz_A      ┆        ┆        ┆   ┆      ┆               ┆       ┆           │
│ |330…          ┆                ┆        ┆        ┆   ┆      ┆               ┆       ┆           │
│ IMGVR_UViG_330 ┆ 6yfh-assembly1 ┆ 0.445  ┆ 142    ┆ … ┆ 156  ┆ 6YFH          ┆ AA    ┆ true      │
│ 0043810_000065 ┆ .cif.gz_AA     ┆        ┆        ┆   ┆      ┆               ┆       ┆           │
│ |330…          ┆                ┆        ┆        ┆   ┆      ┆               ┆       ┆           │
└────────────────┴────────────────┴────────┴────────┴───┴──────┴───────────────┴───────┴───────────┘

Now we compute for each protein cluster only containing only unannotated proteins and VOG-detectable capsids the proportion of unannotated proteins that have structural similarity to capsids in the above table:

In [10]:
results: list[pl.LazyFrame] = []

ptn_embedding_methods = [
    "esm-small", "esm-large", "pst-small", "pst-large", "genslm"

]
with tb.open_file("embedding-based_protein_clusters_per_genome_cluster.h5") as fp:
    for gm_node in fp.root:
        genome_method = gm_node._v_name

        for ptn_method in ptn_embedding_methods:
            clustering_metadata = gm_node["metadata"][:]
            ptn_clusters = gm_node[ptn_method][:]
            genome_clusters = gm_node["genome"][:]

            for pclu, gclu, (genome_k, genome_res, protein_k, protein_res) in zip(
                ptn_clusters, genome_clusters, clustering_metadata
            ):
                if (genome_k != 15) or (genome_res != 1.0) or (protein_res == 0.85):
                    continue

                cluster_df = (
                    ptn_info
                    .lazy()
                    .with_columns(
                        genome_cluster = pl.lit(gclu),
                        protein_cluster = pl.lit(pclu),
                    )
                    .with_columns(
                        cluster_size = pl.len().over("genome_cluster", "protein_cluster")
                    )
                    .filter(pl.col("cluster_size") > 1)
                    .join(struct_aln.lazy(), left_on="ptn", right_on="query")
                    .with_columns(
                        ptn_method=pl.lit(ptn_method),
                        genome_method=pl.lit(genome_method),
                        genome_k=pl.lit(genome_k),
                        genome_r=pl.lit(genome_res),
                        ptn_r=pl.lit(protein_res),
                        ptn_k=pl.lit(protein_k),
                    )
                )

                results.append(cluster_df)

cluster_summary = pl.concat(pl.collect_all(results))
cluster_summary

shape: (10_461_674, 30)
┌───────────────┬─────────┬──────────────┬───────────────┬───┬──────────┬──────────┬───────┬───────┐
│ ptn           ┆ ptn_id  ┆ vog_bitscore ┆ vog_annot     ┆ … ┆ genome_k ┆ genome_r ┆ ptn_r ┆ ptn_k │
│ ---           ┆ ---     ┆ ---          ┆ ---           ┆   ┆ ---      ┆ ---      ┆ ---   ┆ ---   │
│ str           ┆ i64     ┆ f64          ┆ str           ┆   ┆ i32      ┆ f64      ┆ f64   ┆ i32   │
╞═══════════════╪═════════╪══════════════╪═══════════════╪═══╪══════════╪══════════╪═══════╪═══════╡
│ IMGVR_UViG_25 ┆ 43      ┆ null         ┆ unknown       ┆ … ┆ 15       ┆ 1.0      ┆ 0.1   ┆ 5     │
│ 79778745_0000 ┆         ┆              ┆ function      ┆   ┆          ┆          ┆       ┆       │
│ 01|257…       ┆         ┆              ┆               ┆   ┆          ┆          ┆       ┆       │
│ IMGVR_UViG_26 ┆ 1012    ┆ 260.42       ┆ sp|P15895|ORF ┆ … ┆ 15       ┆ 1.0      ┆ 0.1   ┆ 5     │
│ 48501348_0000 ┆         ┆              ┆ 4_SPV1R       ┆   ┆          ┆          ┆       ┆       │
│ 05|264…       ┆         ┆              ┆ Uncharacter…  ┆   ┆          ┆          ┆       ┆       │
│ IMGVR_UViG_26 ┆ 1014    ┆ null         ┆ unknown       ┆ … ┆ 15       ┆ 1.0      ┆ 0.1   ┆ 5     │
│ 48501348_0000 ┆         ┆              ┆ function      ┆   ┆          ┆          ┆       ┆       │
│ 05|264…       ┆         ┆              ┆               ┆   ┆          ┆          ┆       ┆       │
│ IMGVR_UViG_26 ┆ 1016    ┆ 332.97       ┆ sp|P15893|ORF ┆ … ┆ 15       ┆ 1.0      ┆ 0.1   ┆ 5     │
│ 48501348_0000 ┆         ┆              ┆ 2_SPV1R       ┆   ┆          ┆          ┆       ┆       │
│ 05|264…       ┆         ┆              ┆ Uncharacter…  ┆   ┆          ┆          ┆       ┆       │
│ IMGVR_UViG_26 ┆ 1021    ┆ null         ┆ unknown       ┆ … ┆ 15       ┆ 1.0      ┆ 0.1   ┆ 5     │
│ 48501348_0000 ┆         ┆              ┆ function      ┆   ┆          ┆          ┆       ┆       │
│ 05|264…       ┆         ┆              ┆               ┆   ┆          ┆          ┆       ┆       │
│ …             ┆ …       ┆ …            ┆ …             ┆ … ┆ …        ┆ …        ┆ …     ┆ …     │
│ IMGVR_UViG_GV ┆ 7181768 ┆ 338.24       ┆ sp|A7U6E9|MCP ┆ … ┆ 15       ┆ 1.0      ┆ 0.5   ┆ 15    │
│ MAG-S-ERX5561 ┆         ┆              ┆ _POV01 Major  ┆   ┆          ┆          ┆       ┆       │
│ 27-47_…       ┆         ┆              ┆ capsid…       ┆   ┆          ┆          ┆       ┆       │
│ IMGVR_UViG_GV ┆ 7181857 ┆ null         ┆ unknown       ┆ … ┆ 15       ┆ 1.0      ┆ 0.5   ┆ 15    │
│ MAG-S-ERX5561 ┆         ┆              ┆ function      ┆   ┆          ┆          ┆       ┆       │
│ 33-39_…       ┆         ┆              ┆               ┆   ┆          ┆          ┆       ┆       │
│ IMGVR_UViG_GV ┆ 7182107 ┆ null         ┆ unknown       ┆ … ┆ 15       ┆ 1.0      ┆ 0.5   ┆ 15    │
│ MAG-S-ERX5561 ┆         ┆              ┆ function      ┆   ┆          ┆          ┆       ┆       │
│ 33-39_…       ┆         ┆              ┆               ┆   ┆          ┆          ┆       ┆       │
│ IMGVR_UViG_GV ┆ 7182117 ┆ 659.96       ┆ sp|A7U6E9|MCP ┆ … ┆ 15       ┆ 1.0      ┆ 0.5   ┆ 15    │
│ MAG-S-ERX5561 ┆         ┆              ┆ _POV01 Major  ┆   ┆          ┆          ┆       ┆       │
│ 33-39_…       ┆         ┆              ┆ capsid…       ┆   ┆          ┆          ┆       ┆       │
│ IMGVR_UViG_GV ┆ 7182202 ┆ 267.94       ┆ sp|A7U6E9|MCP ┆ … ┆ 15       ┆ 1.0      ┆ 0.5   ┆ 15    │
│ MAG-S-ERX5561 ┆         ┆              ┆ _POV01 Major  ┆   ┆          ┆          ┆       ┆       │
│ 33-39_…       ┆         ┆              ┆ capsid…       ┆   ┆          ┆          ┆       ┆       │
└───────────────┴─────────┴──────────────┴───────────────┴───┴──────────┴──────────┴───────┴───────┘

In [11]:
group_cols = ["genome_method", "ptn_method", "genome_k", "genome_r", "ptn_k", "ptn_r"]

# need this to compute prop of only unannotated proteins that are capsids
detected_capsids = (
    cluster_summary
    .filter(pl.col("vog_category") != "unknown")
    .group_by(*group_cols, "genome_cluster", "protein_cluster")
    .agg(
        known_mcps = pl.len()
    )
)

detected_capsids

shape: (5_856_744, 9)
┌────────────┬────────────┬──────────┬──────────┬───┬───────┬────────────┬────────────┬────────────┐
│ genome_met ┆ ptn_method ┆ genome_k ┆ genome_r ┆ … ┆ ptn_r ┆ genome_clu ┆ protein_cl ┆ known_mcps │
│ hod        ┆ ---        ┆ ---      ┆ ---      ┆   ┆ ---   ┆ ster       ┆ uster      ┆ ---        │
│ ---        ┆ str        ┆ i32      ┆ f64      ┆   ┆ f64   ┆ ---        ┆ ---        ┆ u32        │
│ str        ┆            ┆          ┆          ┆   ┆       ┆ u32        ┆ u32        ┆            │
╞════════════╪════════════╪══════════╪══════════╪═══╪═══════╪════════════╪════════════╪════════════╡
│ ctx-avg-la ┆ esm-small  ┆ 15       ┆ 1.0      ┆ … ┆ 0.1   ┆ 480        ┆ 0          ┆ 1          │
│ rge        ┆            ┆          ┆          ┆   ┆       ┆            ┆            ┆            │
│ ctx-avg-la ┆ esm-small  ┆ 15       ┆ 1.0      ┆ … ┆ 0.1   ┆ 855        ┆ 3          ┆ 2          │
│ rge        ┆            ┆          ┆          ┆   ┆       ┆            ┆            ┆            │
│ ctx-avg-la ┆ esm-small  ┆ 15       ┆ 1.0      ┆ … ┆ 0.1   ┆ 1290       ┆ 2          ┆ 2          │
│ rge        ┆            ┆          ┆          ┆   ┆       ┆            ┆            ┆            │
│ ctx-avg-la ┆ esm-small  ┆ 15       ┆ 1.0      ┆ … ┆ 0.1   ┆ 2058       ┆ 1          ┆ 2          │
│ rge        ┆            ┆          ┆          ┆   ┆       ┆            ┆            ┆            │
│ ctx-avg-la ┆ esm-small  ┆ 15       ┆ 1.0      ┆ … ┆ 0.1   ┆ 455        ┆ 3          ┆ 3          │
│ rge        ┆            ┆          ┆          ┆   ┆       ┆            ┆            ┆            │
│ …          ┆ …          ┆ …        ┆ …        ┆ … ┆ …     ┆ …          ┆ …          ┆ …          │
│ pst-small  ┆ genslm     ┆ 15       ┆ 1.0      ┆ … ┆ 0.5   ┆ 64254      ┆ 74         ┆ 1          │
│ pst-small  ┆ genslm     ┆ 15       ┆ 1.0      ┆ … ┆ 0.5   ┆ 64069      ┆ 356        ┆ 1          │
│ pst-small  ┆ genslm     ┆ 15       ┆ 1.0      ┆ … ┆ 0.5   ┆ 64055      ┆ 455        ┆ 1          │
│ pst-small  ┆ genslm     ┆ 15       ┆ 1.0      ┆ … ┆ 0.5   ┆ 64145      ┆ 4          ┆ 1          │
│ pst-small  ┆ genslm     ┆ 15       ┆ 1.0      ┆ … ┆ 0.5   ┆ 64315      ┆ 75         ┆ 1          │
└────────────┴────────────┴──────────┴──────────┴───┴───────┴────────────┴────────────┴────────────┘

In [12]:
per_cluster_proportion_summary = (
    cluster_summary
    .filter(
        pl.col("is_capsid") & (pl.col("vog_category") == "unknown")
    )
    .group_by(*group_cols, "genome_cluster", "protein_cluster")
    .agg(
        pl.first("cluster_size"),
        capsid_hits = pl.len(),
    )
    .join(detected_capsids, on=group_cols + ["genome_cluster", "protein_cluster"])
    .with_columns(
        prop = pl.col("capsid_hits") / (pl.col("cluster_size") - pl.col("known_mcps")),
        weight = pl.col("cluster_size") / pl.sum("cluster_size").over(group_cols)
    )
    .with_columns(
        weighted_prop = pl.col("prop") * pl.col("weight")
    )
)

per_cluster_proportion_summary

shape: (247_191, 14)
┌────────────┬────────────┬──────────┬──────────┬───┬────────────┬──────────┬──────────┬───────────┐
│ genome_met ┆ ptn_method ┆ genome_k ┆ genome_r ┆ … ┆ known_mcps ┆ prop     ┆ weight   ┆ weighted_ │
│ hod        ┆ ---        ┆ ---      ┆ ---      ┆   ┆ ---        ┆ ---      ┆ ---      ┆ prop      │
│ ---        ┆ str        ┆ i32      ┆ f64      ┆   ┆ u32        ┆ f64      ┆ f64      ┆ ---       │
│ str        ┆            ┆          ┆          ┆   ┆            ┆          ┆          ┆ f64       │
╞════════════╪════════════╪══════════╪══════════╪═══╪════════════╪══════════╪══════════╪═══════════╡
│ ctx-avg-la ┆ esm-small  ┆ 15       ┆ 1.0      ┆ … ┆ 1          ┆ 0.090909 ┆ 0.000462 ┆ 0.000042  │
│ rge        ┆            ┆          ┆          ┆   ┆            ┆          ┆          ┆           │
│ ctx-avg-la ┆ esm-small  ┆ 15       ┆ 1.0      ┆ … ┆ 2          ┆ 0.090909 ┆ 0.0005   ┆ 0.000045  │
│ rge        ┆            ┆          ┆          ┆   ┆            ┆          ┆          ┆           │
│ ctx-avg-la ┆ esm-small  ┆ 15       ┆ 1.0      ┆ … ┆ 1          ┆ 0.058824 ┆ 0.000693 ┆ 0.000041  │
│ rge        ┆            ┆          ┆          ┆   ┆            ┆          ┆          ┆           │
│ ctx-avg-la ┆ esm-small  ┆ 15       ┆ 1.0      ┆ … ┆ 1          ┆ 0.181818 ┆ 0.000462 ┆ 0.000084  │
│ rge        ┆            ┆          ┆          ┆   ┆            ┆          ┆          ┆           │
│ ctx-avg-la ┆ esm-small  ┆ 15       ┆ 1.0      ┆ … ┆ 1          ┆ 0.05     ┆ 0.000808 ┆ 0.00004   │
│ rge        ┆            ┆          ┆          ┆   ┆            ┆          ┆          ┆           │
│ …          ┆ …          ┆ …        ┆ …        ┆ … ┆ …          ┆ …        ┆ …        ┆ …         │
│ pst-small  ┆ genslm     ┆ 15       ┆ 1.0      ┆ … ┆ 1          ┆ 0.117647 ┆ 0.000626 ┆ 0.000074  │
│ pst-small  ┆ genslm     ┆ 15       ┆ 1.0      ┆ … ┆ 2          ┆ 0.045455 ┆ 0.000835 ┆ 0.000038  │
│ pst-small  ┆ genslm     ┆ 15       ┆ 1.0      ┆ … ┆ 1          ┆ 0.083333 ┆ 0.000452 ┆ 0.000038  │
│ pst-small  ┆ genslm     ┆ 15       ┆ 1.0      ┆ … ┆ 1          ┆ 0.142857 ┆ 0.000278 ┆ 0.00004   │
│ pst-small  ┆ genslm     ┆ 15       ┆ 1.0      ┆ … ┆ 1          ┆ 0.066667 ┆ 0.000557 ┆ 0.000037  │
└────────────┴────────────┴──────────┴──────────┴───┴────────────┴──────────┴──────────┴───────────┘

Then, we use a weighted mean (weights = protein cluster size) to summarize for each genome clustering / protein clustering configuration:

In [13]:
proportion_summary = (
    per_cluster_proportion_summary
    .group_by(group_cols)
    .agg(
        pl.sum("weighted_prop", "cluster_size")
    )
    .rename({"weighted_prop": "prop"})
    .sort(["prop", "cluster_size"], descending=True)
)

proportion_summary

shape: (180, 8)
┌───────────────┬────────────┬──────────┬──────────┬───────┬───────┬──────────┬──────────────┐
│ genome_method ┆ ptn_method ┆ genome_k ┆ genome_r ┆ ptn_k ┆ ptn_r ┆ prop     ┆ cluster_size │
│ ---           ┆ ---        ┆ ---      ┆ ---      ┆ ---   ┆ ---   ┆ ---      ┆ ---          │
│ str           ┆ str        ┆ i32      ┆ f64      ┆ i32   ┆ f64   ┆ f64      ┆ u32          │
╞═══════════════╪════════════╪══════════╪══════════╪═══════╪═══════╪══════════╪══════════════╡
│ genslm        ┆ pst-small  ┆ 15       ┆ 1.0      ┆ 5     ┆ 0.5   ┆ 0.387778 ┆ 3299         │
│ ctx-avg-large ┆ pst-small  ┆ 15       ┆ 1.0      ┆ 5     ┆ 0.5   ┆ 0.376659 ┆ 3062         │
│ esm2_t30_150M ┆ pst-small  ┆ 15       ┆ 1.0      ┆ 5     ┆ 0.5   ┆ 0.374475 ┆ 2921         │
│ pst-small     ┆ pst-small  ┆ 15       ┆ 1.0      ┆ 5     ┆ 0.5   ┆ 0.371701 ┆ 2875         │
│ hyena-dna     ┆ pst-small  ┆ 15       ┆ 1.0      ┆ 5     ┆ 0.5   ┆ 0.370921 ┆ 3202         │
│ …             ┆ …          ┆ …        ┆ …        ┆ …     ┆ …     ┆ …        ┆ …            │
│ esm2_t6_8M    ┆ esm-small  ┆ 15       ┆ 1.0      ┆ 15    ┆ 0.1   ┆ 0.026478 ┆ 105365       │
│ esm2_t30_150M ┆ esm-small  ┆ 15       ┆ 1.0      ┆ 15    ┆ 0.1   ┆ 0.025885 ┆ 104099       │
│ ctx-avg-large ┆ esm-small  ┆ 15       ┆ 1.0      ┆ 15    ┆ 0.1   ┆ 0.024675 ┆ 119255       │
│ ctx-avg-small ┆ esm-large  ┆ 15       ┆ 1.0      ┆ 15    ┆ 0.1   ┆ 0.023028 ┆ 64637        │
│ ctx-avg-small ┆ esm-small  ┆ 15       ┆ 1.0      ┆ 15    ┆ 0.1   ┆ 0.021987 ┆ 69974        │
└───────────────┴────────────┴──────────┴──────────┴───────┴───────┴──────────┴──────────────┘

The above table was used to make **Extended Data Figure 9A**.

For **Figure 4C**, we further summarized this to focus on the differences between protein embeddings used for clustering, regardless of genome clustering:

In [14]:
(
    proportion_summary
    .group_by("ptn_method", "ptn_k", "ptn_r")
    .agg(
        pl.mean("prop")
    )
    .sort(["ptn_k", "ptn_r", "prop"], descending=[False, False, True])
)

shape: (20, 4)
┌────────────┬───────┬───────┬──────────┐
│ ptn_method ┆ ptn_k ┆ ptn_r ┆ prop     │
│ ---        ┆ ---   ┆ ---   ┆ ---      │
│ str        ┆ i32   ┆ f64   ┆ f64      │
╞════════════╪═══════╪═══════╪══════════╡
│ pst-small  ┆ 5     ┆ 0.1   ┆ 0.141046 │
│ pst-large  ┆ 5     ┆ 0.1   ┆ 0.105428 │
│ genslm     ┆ 5     ┆ 0.1   ┆ 0.084609 │
│ esm-small  ┆ 5     ┆ 0.1   ┆ 0.074845 │
│ esm-large  ┆ 5     ┆ 0.1   ┆ 0.073542 │
│ …          ┆ …     ┆ …     ┆ …        │
│ pst-small  ┆ 15    ┆ 0.5   ┆ 0.221303 │
│ pst-large  ┆ 15    ┆ 0.5   ┆ 0.091942 │
│ genslm     ┆ 15    ┆ 0.5   ┆ 0.087665 │
│ esm-small  ┆ 15    ┆ 0.5   ┆ 0.079305 │
│ esm-large  ┆ 15    ┆ 0.5   ┆ 0.077922 │
└────────────┴───────┴───────┴──────────┘